File: AFView_old/Pod/Categories/UIView+FirstResponder.m
#import "UIView+FirstResponder.h"


#pragma mark Class Definition

@implementation UIView (FirstResponder)


#pragma mark - Public Methods

+ (UIView *)findFirstResponder
{
	UIView *mainView = [UIApplication sharedApplication].keyWindow;
	return [mainView findFirstResponder];
}

+ (void)resignFirstResponderWithCompletion: (void (^)(void))completion
{
	UIView *firstResponder = [UIView findFirstResponder];
	
	if (firstResponder != nil)
	{
		[firstResponder resignFirstResponder];
		
		// Call completion, after a delay.
		if (completion != nil)
		{
			// Copy the block;
			completion = [completion copy];
			
			// TODO: CLUDGE: Call the block after a delay.
			// Perhaps use a keyboard observer if the keyboard is up...
			[self performSelector: @selector(_performBlock:)
				withObject: completion
				afterDelay: 0.3];
		}
	}
	else
	{
		// Call completion, if provided.
		if (completion != nil)
		{
			completion();
		}
	}
}

- (UIView *)findFirstResponder
{
    if (self.isFirstResponder)
	{
        return self;     
    }

    for (UIView *subView in self.subviews)
	{
        UIView *firstResponder = [subView findFirstResponder];
		
        if (firstResponder != nil)
		{
            return firstResponder;
        }
    }

    return nil;
}


#pragma mark - Private Methods

+ (void)_performBlock: (void (^)(void))block
{
	block();
}

@end

File: AFView/Pod/Categories/UIView+FirstResponder.swift
import Foundation

extension UIView
{
	// MARK: - Type Aliases
	
	public typealias UIViewResignationBlock = (Void) -> Void


	// MARK: - Public Methods

	public static func findFirstResponder()
		-> UIView?
	{
		let mainView: UIView? = UIApplication.sharedApplication().keyWindow
		return mainView?.findFirstResponder()
	}
	
	public static func resignFirstResponder(completion: UIViewResignationBlock)
	{
		let firstResponder: UIView? = UIView.findFirstResponder()
		
		if firstResponder != nil
		{
			firstResponder?.resignFirstResponder()
			
			// CLUDGE: Call the block after a delay.
			_delay(0.3, closure: completion)
		}
		else
		{
			// Call completion, if provided.
			completion()
		}
	}
	
	public func findFirstResponder()
		-> UIView?
	{
		// Return once the first responder is found.
		if self.isFirstResponder()
		{
			return self
		}
		
		// Recursively search subviews.
		for subview in self.subviews
		{
			let firstResponder: UIView? = subview.findFirstResponder()
			
			if firstResponder != nil
			{
				return firstResponder
			}
		}
		
		return nil
	}
	
	// MARK: - Private Methods
	
	private static func _delay(delay: Double, closure: ()->())
	{
		dispatch_after(
			dispatch_time(
				DISPATCH_TIME_NOW,
				Int64(delay * Double(NSEC_PER_SEC))
			),
			dispatch_get_main_queue(),
			closure)
	}
}